### ISSUE TITLE
GPUParticles3D inherit velocity is broken

### ISSUE THREAD

**Issue by huwpascoe** [godotengine/godot#97160]
### Tested versions

4.3

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 960 (NVIDIA; 32.0.15.6081) - Intel(R) Core(TM) i5-6600K CPU @ 3.50GHz (4 Threads)

### Issue description

inherit_velocity_ratio doesn't work as expected. Particles seem to emit ahead of where they're supposed to and this gap increases the higher the velocity.

The particle effect is placed at the back of the rocket, yet at this velocity they're being emitted near the front!
![image](https://github.com/user-attachments/assets/5e9dd128-0bbe-4508-ab0d-3882209f98ad)

### Steps to reproduce

Set inherit_velocity_ratio to greater than zero.

### Minimal reproduction project (MRP)

[particles_velocity_ratio.zip](https://github.com/user-attachments/files/17049711/particles_velocity_ratio.zip)
![image](https://github.com/user-attachments/assets/a0765229-b284-464f-9a6b-d9d1bd0c4ee6)

**Comment by CycloneRing** [comment]
Same issue

**Comment by KingTheFifth** [comment]
Hiya! I am planning to look into this, just to give people a heads up

**Comment by Calinou** [comment]
Can you reproduce this in `master`? https://github.com/godotengine/godot/pull/93837 was merged there but hasn't been cherry-picked to 4.3.x yet.

**Comment by huwpascoe** [comment]
Confirmed in rendering meeting that no, the issue is not technically resolved. It should work for all the settings. Please go ahead @KingTheFifth

**Comment by KingTheFifth** [comment]
Ok, thanks for confirming! I will begin working on it now

**Comment by JackErb** [comment]
Any update on this? :)

**Comment by KingTheFifth** [comment]
Hi! Sorry for the late reply. I have not managed to make any significant progress. If anything, it seems that inherit velocity does not even work as expected with `fixed_fps=60`, `interpolate=false` and `fract_delta=false`. Running the MPR with inherit velocity ratio set to 1.0 on commit 33e1f570ff4c437ad336085a1f605e0d2705ef54 produces the following:

https://github.com/user-attachments/assets/057ac579-1962-448a-9f5d-85688ed96be7

With inherit velocity ratio 1.0, I expected the particles to remain static relative to the rocket as if the rocket was not moving. From the video it should be apparent that the particles gain velocity such that their 'spawn point' move from the left end of the rocket toward the right end. For reference, this is what it looks like if the rocket does not move i.e. `accel = 0.0` in `rocket.gd`.

https://github.com/user-attachments/assets/3791d0a0-1593-4f06-88d3-ca1c6174e58f

I think this issue may be too advanced for me, so anyone with more experience of the physics part of Godot than me is more than welcome to take a look.

**Comment by huwpascoe** [comment]
It is a difficult one to solve, thanks for trying.

**Comment by JackErb** [comment]
I can confirm what @KingTheFifth said that this is still broken with `fixed_fps=60`, `interpolate=false` and `fract_delta=false` on commit https://github.com/godotengine/godot/commit/33e1f570ff4c437ad336085a1f605e0d2705ef54

I've discovered that the problem is the particles are one frame "ahead" of where the object is. I fixed this locally by not applying the inherited velocity on the first frame of particle creation. This works great while `fixed_fps == actual_runtime_fps`, but is broken in other cases such as fixed_fps being lower/higher, or if the actual runtime fps is not stable.

I am going to continue investigating the fix to make this work when then actual runtime fps does not matched the fixed fps

**Comment by JackErb** [comment]
After my investigation, I found we must simulate the particle at the exact tick rate as the physics simulation in order to keep it synchronized across all cases. This is incompatible with the existing option `fixed_fps`. I made a new checkbox for this behavior, named `use_physics_step` (open to feedback on the name).

See Pull Request GH-100453

**Comment by QbieShay** [comment]
Hey 👋 

I can see that there's already been a lot of work in fixing this, but after seeing the complexity added by @JackErb 's PR, I need to come back to this and ask how critical this fix is. I am not new to having limitations on what the particle system can do, so before we go add functions in core, I want to know how critical it is that we can make a rocket look like that by using the specific feature of inherit velocity. 

I would like if you all can take some time to try and get acceptable results by combining multiple particle systems and things of the sort.

Lastly JackErb, thank you a lot for investing time in this. I know it sucks I'm inviting everyone to go back to square one, but it's because I see the complexity of your work that as a maintainer I have to go back and ask, is this complexity really necessary. Nonetheless I appreciate the work and the commitment. Thank you.

**Comment by huwpascoe** [comment]
> need to come back to this and ask how critical this fix is.

It's not Game Over yet! What's important is JackErb has a working prototype. Glancing at the PR, I've got a hunch that complexity can be reduced...

**Comment by QbieShay** [comment]
Well, I would still like that the initial question is answered: can we obtain a good enough result without radically overhauling the way particles are processed? Why does this rocket 's smoke need to be done with inherit velocity? Is there another way?

**Comment by JackErb** [comment]
In my space based "orbital mechanics" game this issue was critical enough for me to spend several weeks solving this issue. I could not achieve the desired effect without using inherit velocity. If inherit velocity is disabled then particles would not behave as realistically expected on fast moving objects, they would be quickly whisked off screen when my object was traveling at high speeds which ruined the visual fidelity of the particles I had created (since I can no longer see them!) and created a disconnect between the expected and observed physical simulation. Note that the fidelity of the physical simulation is important in my game as it is about realistically manipulating orbits and realistic space flight (think Kerbal Space Program as an analogy).

The closest version of this NOT using inherit velocity would be to spin up my own particle system using nodes which loses all the handy features of the ParticleProcessMaterial and has significant performance implications.

I will certainly be using my fix locally if it is not picked up here. :)

**Comment by QbieShay** [comment]
Did you try to use particles simulated in local space?

**Comment by QbieShay** [comment]
And frankly I do wonder if it wouldn't be better to do the magical "I'll eventually schedule this" port of process material to CPU. That way processing on the physics tick would be a zero effort change

**Comment by QbieShay** [comment]
So, I think in general  there will always be edge cases where things don't quite work out. I think it's very normal in game development.

I think also that the most "correct" solution isn't always the best one. I understand that you'd like your usecase to be supported in a built-in manner, but this could just be one of such cases where you'd use a custom particle process shader. I  know it's not ideal nor convenient, but I think in order to figure out which changes we need to do to the engine, we also need to fully understand the limitations and if there are workarounds. In particular considering what i mentioned about porting process material to CPU. I think that'd be the best solution that would catch a bunch of other corner cases and UX issues that people have already mentioned.

I believe we need a bit more investigation before committing to a solution. @huwpascoe also mentioned in RC the possibility of syncing the fixed fps with the physics process. Perhaps we can just have a setting to skip applying velocity in the first frame of a particle system as a render mode.

**Comment by JackErb** [comment]
Note that setting the particle's fixed fps to the same as the physics tick rate fixes the issue but only while the framerate is stable. I still observed substantial desync when the machine framerate gets too low. Which is why I went with the full version in my PR

I could see the CPUParticles being a good avenue to go down

**Comment by huwpascoe** [comment]
Yeah if you drag the app window around (on Windows systems), then you can see the desync in action as it plays catch up with the message queue.

**Comment by jonpecar** [comment]
I personally would also like to see this issue resolved. I have similar issues for similar reasons (orbital mechanic space game) and neither the CPU or the GPU particles really provide a valid solution to the problem. In addition to the CPUParticles missing the features of the Process Material, they also do not have the amount ratio property that allows GPU particles to be dynamically adjusted in quantity.

Given the state my game is currently at I am hoping that a fix for this will be implemented into either particle system. Otherwise plan B would be to do something with nodes as particles or another custom solution as has already been discussed.

I do think that the approach of using the physics step would be very valuable and I suspect that this use case is probably not actually that niche as it would apply to many space games using particles for thrust etc.

**Comment by QbieShay** [comment]
CPUParticles on being on par with GPUParticles is a different issue: this could be solved by updating cpuparticles to have the same features as gpuparticles and adding amount_ratio

**Comment by QbieShay** [comment]
For the time being I suggest that whoever needs this feature writes their own cpuparticles with multimeshes (strictly for the particles that need to stay synched to their physics bodies in that way). I honestly do not think that it's a massive endeavour, compared to what it may cost to everyone using godot to sync GPUParticle process with physics process. 

NOTE: I'm not proposing this as a final solution. I think eventually support for this could come natively in Godot, but for now the solution of syncing gpu with physics process doesn't sound like the best option in the long run.

**Comment by jonpecar** [comment]
I think if CPU particles had feature parity with GPU particles and could synchronize with the physics step that would provide a reasonable solution especially where a high particle count is not required. It would definitely suffice in my situation.

That said, in the interim I will, as you suggest, implement my own CPU particles.

**Comment by QbieShay** [comment]
Changed from bug to enhancement as a follow up of the previous discussion, since the solution to this is to reach feature parity between gpuparticles and cpuparticles
